unittests/Feature/FeatureModelBuilder.cpp

```diff
@@ -157,47 +160,102 @@ TEST(FeatureModelBuilder, detectXMLAlternativesBroken) {
+TEST(FeatureModelBuilder, detectXMLAlternativesOptionalBroken) {
   FeatureModelBuilder B;
   B.makeFeature<BinaryFeature>("a");
   B.addEdge("a", "aa");
   B.addEdge("a", "ab");
+  B.makeFeature<BinaryFeature>("aa", false);
+  B.makeFeature<BinaryFeature>("ab", true);
+
+  B.addConstraint(std::make_unique<ExcludesConstraint>(
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("aa")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ab"))));
+  B.addConstraint(std::make_unique<ExcludesConstraint>(
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ab")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("aa"))));
+  auto FM = B.buildFeatureModel();
+  assert(FM);
+
+  EXPECT_TRUE(FM->getFeature("a")->hasEdgeTo(*FM->getFeature("aa")));
+  EXPECT_TRUE(FM->getFeature("a")->hasEdgeTo(*FM->getFeature("ab")));
+}
+
+TEST(FeatureModelBuilder, detectXMLAlternativesOutOfOrder) {
+  FeatureModelBuilder B;
+  B.makeFeature<BinaryFeature>("a");
+  B.addEdge("a", "ab");
+  B.addEdge("a", "aa");
   B.addEdge("a", "ac");
-  B.addEdge("a", "ad");
-  B.addEdge("a", "ae");
   B.makeFeature<BinaryFeature>("aa", false);
   B.makeFeature<BinaryFeature>("ac", false);
-  B.makeFeature<BinaryFeature>("ae", false);
-  auto *AB = B.makeFeature<BinaryFeature>("ab", false);
-  auto *AD = B.makeFeature<BinaryFeature>("ad", false);
+  B.makeFeature<BinaryFeature>("ab", false);
 
   B.addConstraint(std::make_unique<ExcludesConstraint>(
-      std::make_unique<PrimaryFeatureConstraint>(AB),
-      std::make_unique<PrimaryFeatureConstraint>(AD)));
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("aa")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ab"))));
+  B.addConstraint(std::make_unique<ExcludesConstraint>(
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ab")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("aa"))));
+  B.addConstraint(std::make_unique<ExcludesConstraint>(
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ac")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("aa"))));
+  B.addConstraint(std::make_unique<ExcludesConstraint>(
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ab")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ac"))));
+  B.addConstraint(std::make_unique<ExcludesConstraint>(
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("aa")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ac"))));
```

Comment:
This pattern looks repetitive, I think we should create a helper/short cut for this.

Reply:
I added a macro. 
With this the count of `std::` is significantly reduced, so I think we do not need to add `using namespace std` here.

Reply:
Please don't use macros :cry:, it should be possible to just build is as a constexpr function or maybe a template func.